tfhe/src/integer/server_key/radix_parallel/div_mod.rs

```diff
@@ -113,134 +114,241 @@ impl ServerKey {
+            // to trim out bits which should not be there
+
+            let mut trim_last_interesting_divisor_bits = || {
+                if ((msb_bit_set + 1) % num_bits_in_message as usize) == 0 {
+                    return;
+                }
+                // The last block of the interesting part of the remainder
+                // can contain bits which we should not account for
+                // we have to zero them out.
+
+                // Where the msb is set in the block
+                let pos_in_block = msb_bit_set as u64 % num_bits_in_message;
+
+                // e.g 2 bits in message:
+                // if pos_in_block is 0, then we want to keep only first bit (right shift mask
+                // by 1) if pos_in_block is 1, then we want to keep the two
+                // bits (right shift mask by 0)
+                let shift_amount = num_bits_in_message - (pos_in_block + 1);
+                // Create mask of 1s on the message part, 0s in the carries
+                let full_message_mask = self.key.message_modulus.0 as u64 - 1;
+                // Shift the mask so that we will only keep bits we should
+                let shifted_mask = full_message_mask >> shift_amount;
+
+                let masking_lut = self.key.generate_lookup_table(|x| x & shifted_mask);
+                self.key.apply_lookup_table_assign(
+                    interesting_divisor.blocks.last_mut().unwrap(),
+                    &masking_lut,
+                );
+            };
 
-            // For comparisons, trivial are dealt with differently
-            let (non_trivial_blocks_are_ge, trivial_blocks_are_zero) = rayon::join(
+            let mut trim_first_divisor_ms_bits = || {
+                if divisor_ms_blocks.blocks.is_empty()
+                    || ((msb_bit_set + 1) % num_bits_in_message as usize) == 0
+                {
+                    return;
+                }
+                // As above, we need to zero out some bits, but here its in the
+                // first block of most significant blocks of the divisor.
+                // The block has the same value as the last block of interesting_divisor.
+                // Here we will zero out the bits that the spawn above
+                // wanted to keep.
+
+                // Where the msb is set in the block
+                let pos_in_block = msb_bit_set as u64 % num_bits_in_message;
+
+                // e.g 2 bits in message:
+                // if pos_in_block is 0, then we want to discard the first bit (left shift mask
+                // by 1) if pos_in_block is 1, then we want to discard the
+                // two bits (left shift mask by 2) let shift_amount =
+                // num_bits_in_message - pos_in_block as u64;
+                let shift_amount = pos_in_block + 1;
+                let full_message_mask = self.key.message_modulus.0 as u64 - 1;
+                let shifted_mask = full_message_mask << shift_amount;
```

Comment:
this requires the carry bits to be empty, is it always the case ? I think there is an alternative where you can start with 1, and conditionally AND it with 0 while shifting it by one to only the select the bits you want

Reply:
It does not require empty carry bits, we use a non-bivariate PBS, and the mask computed masks the bits we want, which are some bits of the message part.

Its true tho, that instead of calling applu_lut_assign, I could call bitand_assign